### PR TITLE
Refactored @wordpress/i18n to be an instantiable

### PR DESCRIPTION
**Pull request by jameslnewell**
## Description

As per the proposed solution in https://github.com/WordPress/gutenberg/issues/19210 I've refactored the `@wordpress/i18n` module to expose an instantiable class. e.g.

```js
import { I18N } from '@wordpress/i18n';

const strayaLocale = {
	hello: [ 'gday' ],
};

const frenchLocale = {
	hello: [ 'bonjour' ],
};

const straya = new I18N();
const french = new I18N();

straya.setLocaleData( strayaLocale );
french.setLocaleData( frenchLocale );

expect( straya.__( 'hello' ) ).toEqual( 'gday' );
expect( french.__( 'hello' ) ).toEqual( 'bonjour' );
```

There are no changes to the existing external interface of the module other than the addition of the `I18N` class.

## How has this been tested?

Via the existing suite of unit tests and an additional unit test.

## Screenshots 
N/A

## Types of changes

New feature.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by sirreal** [comment]
🤦‍♂ I just started on this in https://github.com/WordPress/gutenberg/pull/20332

**Comment by sirreal** [comment]
I don't mind which PR moves ahead, this or #20332. However, I suspect the structure #20332 will lead to better tree shaking. Apps that don't use the default instance via importing translate functions (`__`) directly shouldn't pay the cost.

File: packages/i18n/src/index.js [hunk 1]
```diff
@@ -30,12 +20,11 @@ const DEFAULT_LOCALE_DATA = {
  */
 const logErrorOnce = memoize( console.error ); // eslint-disable-line no-console
 
+const i18n = new I18N();
```
Comment (jsnajdr) [inline packages/i18n/src/index.js]: This default instance could also be exported. The default export looks like a good fit.

File: packages/i18n/src/index.js [hunk 2]
```diff
@@ -46,42 +35,8 @@ const i18n = new Tannin( {} );
...
- * @return {string} The translated string.
- */
-function dcnpgettext( domain = 'default', context, single, plural, number ) {
-	if ( ! i18n.data[ domain ] ) {
-		setLocaleData( undefined, domain );
-	}
-
-	return i18n.dcnpgettext( domain, context, single, plural, number );
+export function setLocaleData( data, domain ) {
+	i18n.setLocaleData( data, domain );
```
Comment (jsnajdr) [inline packages/i18n/src/index.js]: This can also be:
```js
export const setLocaleData = i18n.setLocaleData.bind( i18n );
```

Regarding the JSDoc comments: we don't need to duplicate them at two place. If duplicated, they will almost certainly get out of sync sooner or later.

**Comment by jameslnewell** [comment]
> However, I suspect the structure #20332 will lead to better tree shaking. Apps that don't use the default instance via importing translate functions (__) directly shouldn't pay the cost.

@sirreal Are you saying tree shaking won't work unless I use the `.bind(x)` approach? I thought [tree shaking worked fine with the current implementation when only importing the instantiable
 class](https://rollupjs.org/repl/?version=1.31.1&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMCU3QkkxOE4lN0QlMjBmcm9tJTIwJy4lMkZpMThuJyUzQiU1Q24lNUNuY29uc3QlMjBpMThuJTIwJTNEJTIwbmV3JTIwSTE4TigpJTNCJTVDbiU1Q25jb25zb2xlLmxvZyhpMThuKSU1Q24lMjIlMkMlMjJpc0VudHJ5JTIyJTNBdHJ1ZSU3RCUyQyU3QiUyMm5hbWUlMjIlM0ElMjJpMThuLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMiU1Q25leHBvcnQlMjBjbGFzcyUyMEkxOE4lMjAlN0IlNUNuJTVDdF9fKCklMjAlN0IlNUNuJTVDdCU3RCU1Q24lN0QlNUNuJTVDbmNvbnN0JTIwaTE4biUyMCUzRCUyMG5ldyUyMEkxOE4oKSUzQiU1Q24lNUNuZXhwb3J0JTIwZnVuY3Rpb24lMjBfXygpJTIwJTdCJTVDbiU1Q3RyZXR1cm4lMjBpMThuLl9fKCklM0IlNUNuJTdEJTVDbiUyMiU3RCU1RCUyQyUyMm9wdGlvbnMlMjIlM0ElN0IlMjJmb3JtYXQlMjIlM0ElMjJlc20lMjIlMkMlMjJuYW1lJTIyJTNBJTIybXlCdW5kbGUlMjIlMkMlMjJhbWQlMjIlM0ElN0IlMjJpZCUyMiUzQSUyMiUyMiU3RCUyQyUyMmdsb2JhbHMlMjIlM0ElN0IlN0QlN0QlMkMlMjJleGFtcGxlJTIyJTNBbnVsbCU3RA==) ??

File: packages/i18n/src/index.js (hunk 2 above)
Comment (jameslnewell) [reply packages/i18n/src/index.js]: The comments also add type hints so I wouldn't remove them completely. Which copy of the comments would you like me to get rid of? The ones on the class?

**Comment by sirreal** [comment]
> Are you saying tree shaking won't work unless I use the `.bind(x)` approach?

No, I'm saying that if the index only re-exports from other modules, we'll get better tree shaking. As it is `import { I18n } from '@wordpress/i18n` will result in a default instance being instantiated even if its never used. If the default bits move out of the index they become candidates for tree shaking.

This module structure seems ideal (see https://github.com/WordPress/gutenberg/pull/20332):

- index (rexports)
- sprintf
- i18n class
- default i18n instance + exposed methods from instance

File: packages/i18n/src/i18n.js [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+	'': {
+		plural_forms: ( n ) => ( n === 1 ? 0 : 1 ),
+	},
+};
+
+/**
+ * Instantiable I18N methods
+ */
+export class I18N {
+	constructor() {
```
Comment (sirreal) [inline packages/i18n/src/i18n.js]: It would be nice to instatiate the instance with locale data instead of instantiating then calling `setLocaleData`.

File: packages/i18n/src/index.js (hunk 2 above)
Comment (jsnajdr) [reply packages/i18n/src/index.js]: I'm not sure -- but the comments on the class definitions seem to be the canonical ones. The exported auto-bound functions are just helpers.

What kind of type hints do the comments provide? Is there some editor or IDE that can parse the JSDoc comments and provide some autocomplete hints? Like the TypeScript types do?

Comment (sirreal) [reply packages/i18n/src/index.js]: I'm quite confident that TypeScript and most decent tooling will figure out the corret type of `export const setLocaleData = i18n.setLocaleData.bind( i18n );`

However the documentation tooling didn't discover them in #20332. A may be to help it with something like this, which references the canonical type:

```js
/**
 * @typedef {import('./i18n')} I18n
 */
/**
 * @type {I18n['__']}
 */
export const __ = i18n.__.bind( i18n );
/**
 * @type {I18n['_n']}
 */
export const _n = i18n._n.bind( i18n );
//…
```

Comment (jameslnewell) [reply packages/i18n/src/index.js]: That fixes the docs but breaks the typings 😬

**Comment by jameslnewell** [comment]
As per my [example in the rollup repl](https://rollupjs.org/repl/?version=1.31.1&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMCU3QkkxOE4lN0QlMjBmcm9tJTIwJy4lMkZpMThuJyUzQiU1Q24lNUNuY29uc3QlMjBpMThuJTIwJTNEJTIwbmV3JTIwSTE4TigpJTNCJTVDbiU1Q25jb25zb2xlLmxvZyhpMThuKSU1Q24lMjIlMkMlMjJpc0VudHJ5JTIyJTNBdHJ1ZSU3RCUyQyU3QiUyMm5hbWUlMjIlM0ElMjJpMThuLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMiU1Q25leHBvcnQlMjBjbGFzcyUyMEkxOE4lMjAlN0IlNUNuJTVDdF9fKCklMjAlN0IlNUNuJTVDdCU3RCU1Q24lN0QlNUNuJTVDbmNvbnN0JTIwaTE4biUyMCUzRCUyMG5ldyUyMEkxOE4oKSUzQiU1Q24lNUNuZXhwb3J0JTIwZnVuY3Rpb24lMjBfXygpJTIwJTdCJTVDbiU1Q3RyZXR1cm4lMjBpMThuLl9fKCklM0IlNUNuJTdEJTVDbiUyMiU3RCU1RCUyQyUyMm9wdGlvbnMlMjIlM0ElN0IlMjJmb3JtYXQlMjIlM0ElMjJlc20lMjIlMkMlMjJuYW1lJTIyJTNBJTIybXlCdW5kbGUlMjIlMkMlMjJhbWQlMjIlM0ElN0IlMjJpZCUyMiUzQSUyMiUyMiU3RCUyQyUyMmdsb2JhbHMlMjIlM0ElN0IlN0QlN0QlMkMlMjJleGFtcGxlJTIyJTNBbnVsbCU3RA==) I'm pretty confident that the exports are already tree shakable but I've refactored as desired.

**Review by sirreal — COMMENTED** [review]
Overall this looks great, thanks! I've left some notes, but there are no major blockers.

I am concerned about effectively loosing all the documentation for the module. It's important to find a solution for that before this can be released.

File: packages/i18n/src/i18n.js [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+const DEFAULT_LOCALE_DATA = {
+	'': {
+		plural_forms: ( n ) => ( n === 1 ? 0 : 1 ),
+	},
+};
+
+/**
+ * Instantiable I18N methods
+ */
+export class I18N {
```
[the remark this answers is not in the document]
Comment (jsnajdr) [reply packages/i18n/src/i18n.js]: Even if the standards says `I18n`, I think this is a great place to make an exception 🙂 Maybe it's just me, but `I18N` is much nicer to look at.

File: packages/i18n/src/i18n.js [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+	 * @param {LocaleData} [data]   Locale data configuration.
+	 * @param {string}     [domain] Domain for which configuration applies.
+	 */
+	constructor( data, domain ) {
+		/**
+		 * The underlying instance of Tannin to which exported functions interface.
+		 *
+		 * @type {Tannin}
+		 */
+		this.tannin = new Tannin( {}, {} );
```
[the remark this answers is not in the document]
Comment (jameslnewell) [reply packages/i18n/src/i18n.js]: Hmm I added that because the types were complaining... They don't seem to anymore. 🤷‍♂️

**Comment by jameslnewell** [comment]
> I am concerned about effectively loosing all the documentation for the module. It's important to find a solution for that before this can be released.

💯. I think the short term solution is to duplicate the JS docs (or because the tool doesn't document the class anyway, move the documentation to the module methods which do get documented by the tool). Longer term it might be looking at a tool that has a few more smarts e.g. [typedoc](https://typedoc.org/) or [traversing the TS AST](https://github.com/microsoft/TypeScript/wiki/Using-the-Compiler-API#traversing-the-ast-with-a-little-linter).

File: packages/i18n/src/i18n.js [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+const DEFAULT_LOCALE_DATA = {
+	'': {
+		plural_forms: ( n ) => ( n === 1 ? 0 : 1 ),
+	},
+};
+
+/**
+ * Instantiable I18n methods
+ */
+export class I18n {
```
Comment (youknowriad) [inline packages/i18n/src/i18n.js]: I don't mind the class but it seems that the pattern used in other places in Gutenberg is "create" functions instead. Should we make it consistent? (createRegistry for instance)

Comment (jsnajdr) [reply packages/i18n/src/i18n.js]: That's a good observation: the Gutenberg codebase rarely uses `new Something` to instantiate a class. There are factory functions instead.

On the other hand, the `createI18n` factory function is redundant in the sense that it doesn't add anything new: it just forwards the arguments verbatim:
```js
const createI18n = ( ...args ) => new I18n( ...args );
```
And we'll likely need to export the `I18n` class itself, too, one day, because its type is part of the public API.

File: packages/i18n/src/index.js [hunk 7]
```diff
@@ -1,172 +1,7 @@
...
-		return sprintfjs.sprintf( format, ...args );
-	} catch ( error ) {
-		logErrorOnce( 'sprintf error: \n\n' + error.toString() );
-
-		return format;
-	}
-}
+export { sprintf } from './sprintf';
+export { I18n } from './i18n';
+export * from './default-i18n';
```
Comment (jsnajdr) [inline packages/i18n/src/index.js]: I'd personally prefer the more explicit
```js
export { i18n, setLocaleData, __, _x, _n, _nx } from './default-i18n';
```
then, when I want an authoritative answer about what are the exports of the `i18n` module, I only need to look at one file.

File: packages/i18n/src/index.js [hunk 8]
```diff
@@ -1,172 +1,7 @@
 /**
- * External dependencies
+ * Internal dependencies
```
Comment (jsnajdr) [inline packages/i18n/src/index.js]: For a reexport-only module, this docblock doesn't seem to make much sense. If I remove it, the linter doesn't complain.

File: packages/i18n/src/index.js (hunk 2 above)
Comment (jsnajdr) [reply packages/i18n/src/index.js]: If the `docgen` and TypeScript tools are not happy with the bound reexports without jsdoc comment, then I think it's OK to be explicit and re-introduce the duplicate comments.

Would this work better (after adding TS types and the correct JSDoc)?
```js
/**
 * Retrieve the translation of text.
 * @param ...
 * @returns ...
 */
export const __ = ( text, domain ) => i18n.__( text, domain );
```

**Comment by jsnajdr** [comment]
> I've made the changes as desired but as per my example in the rollup repl I'm pretty confident that the exports were already tree shakable and that the refactor won't result in an improvement to what gets outputted.

Rollup and webpack are different bundlers and their tree shaking algorithms are different. The same example won't be tree shaked in webpack.

The Rollup algorithm seems to be more aggressive and it's looking at more things. For example, in the unoptimized program, the `I18N` constructor would be called twice. Rollup looks at the actual constructor and if it's empty, it knows that it can optimize the unused `new I18N` out.

But once I add a `console.log` statement to the constructor, Rollup doesn't optimize and keeps both `new I18N` calls in the bundle!

Webpack's tree shaking is much simpler. It only looks at the `import` and `export` statements and the binding names they use. The actual JS code is mostly invisible to webpack.

**Comment by sirreal** [comment]
> Longer term it might be worth looking at a tool that has a few more smarts […] I think thats well beyond the scope of this PR though.

I agree, it's not trivial: https://github.com/WordPress/gutenberg/issues/18045

The best way for now may simply be to reproduce the types in the index:

```js
/**
 * Merges locale data into the Tannin instance by domain. Accepts data in a
 * Jed-formatted JSON object shape.
 *
 * @see http://messageformat.github.io/Jed/
 *
 * @param {LocaleData} [data]   Locale data configuration.
 * @param {string}     [domain] Domain for which configuration applies.
 */
export { setLocaleData } from './default-i18n';
```

In my testing, that works well enough.

**Comment by sirreal** [comment]
We'll need to rebase this and add the `isRTL` method from #20298.

File: packages/i18n/README.md [hunk 9]
```diff
@@ -27,113 +27,37 @@ For a complete example, see the [Internationalization section of the Block Edito
...
-
--   _single_ `string`: The text to be used if the number is singular.
--   _plural_ `string`: The text to be used if the number is plural.
--   _number_ `number`: The number to compare against to use either the singular or plural form.
--   _domain_ `[string]`: Domain to retrieve the translated text.
-
-_Returns_
-
--   `string`: The translated singular or plural form.
+Undocumented declaration.
```
Comment (aduth) [inline packages/i18n/README.md]: This loss of information is highly problematic to me. I believe that a combination of @jsnajdr 's comment at https://github.com/WordPress/gutenberg/pull/20318/files#r383723605 and an addition of JSDoc to methods in `default-i18n.js` should restore the documentation.

File: packages/list-reusable-blocks/src/components/import-dropdown/index.js [hunk 10]
```diff
@@ -20,11 +20,7 @@ function ImportDropdown( { onUpload } ) {
 			position="bottom right"
 			contentClassName="list-reusable-blocks-import-dropdown__content"
 			renderToggle={ ( { isOpen, onToggle } ) => (
-				<Button
-					aria-expanded={ isOpen }
-					onClick={ onToggle }
-					isPrimary
-				>
+				<Button aria-expanded={ isOpen } onClick={ onToggle } isPrimary>
```
Comment (aduth) [inline packages/list-reusable-blocks/src/components/import-dropdown/index.js]: I don't understand why this change is here?

File: packages/i18n/src/i18n.js (hunk 6 above)
Comment (aduth) [reply packages/i18n/src/i18n.js]: In my own prior thoughts about an implementation, I too had expected this to be implemented as `createI18n`. The note about documenting the type is interesting. We've had reasonable success with documenting object types (e.g. [`createQueue`](https://github.com/WordPress/gutenberg/blob/1b724f316151dd9219905514b0f3401e891eb9ab/packages/priority-queue/src/index.js#L59)), but a class does seem like the more "natively understood" implementation.

I guess it doesn't bother me too much to export this as a class, aside from the inconsistency.

**Comment by aduth** [comment]
> We'll need to rebase this and add the `isRTL` method from #20298.

I think it'd also be nice to use this opportunity to refactor the tests added in #20298 to avoid manipulating the `require` cache and instead create new instances between tests.

These changes:

https://github.com/WordPress/gutenberg/pull/20298/files#diff-acf9737fbf7d5643d202c6452a6dc11eL34-R35

File: packages/i18n/src/index.js (hunk 7 above)
Comment (sirreal) [inline packages/i18n/src/index.js]: We're exporting `i18n` (instance) and `I18n` (class), that differ by the capitalization of the leading `i`. It may be desirable to change one of the names.

Comment (aduth) [reply packages/i18n/src/index.js]: Should we be exporting the default instance at all? 

For prior art, compare to `@wordpress/data`, which is very similar in implementation of having a "default instance", and being able to create separate registries. There are some clear inconsistencies:

&nbsp;|`@wordpress/data`|`@wordpress/i18n`
---|---|---
Instantiate|`createRegistry`|`new I18n`
Default Instance Members|`select`, etc|`__`, etc
Default Instance|_(Not exported)_|`i18n`

File: packages/i18n/src/i18n.js (hunk 6 above)
Comment (youknowriad) [reply packages/i18n/src/i18n.js]: Coming back here, I'm wondering if we should just stay consistent, it doesn't seem like we'll be losing anything by doing so.

File: packages/i18n/src/index.js (hunk 7 above)
Comment (jsnajdr) [reply packages/i18n/src/index.js]: **`createI18n` vs `new I18n`:** The `createI18n()` function feels more like functional programming style, with a regular function that returns a value of certain type. `new I18n` is a classic OOP style. I certainly prefer the former, just keep in mind that in the end the `I18n` name will continue to be one of the package exports. It might not be the executable constructor, but rather the `I18n` type. It's the return type of `createI18n` and can be imported and used:
```js
import { I18n } from '@wordpress/i18n';
```

Similarly, `@wordpress/data` exposes the `DataRegistry` type:
```js
import { DataRegistry, createRegistry } from '@wordpress/data';

const registry: DataRegistry = createRegistry( ... );
```

**Should we be exporting the default instance at all?** I think the question to ask is whether there is any API that takes the whole registry as a parameter and uses it. And the answer is yes. The React bindings use the registry object all the time. `useRegistry` returns it. The `RegistryProvider` passes the registry object as the context value, and `RegistryConsumer` receives it.

The `defaultRegistry` value is the default value of the `RegistryProvider` context. The current implementation [imports it from a private relative path](https://github.com/WordPress/gutenberg/blob/master/packages/data/src/components/registry-provider/context.js#L9).

Imagine that we move the React bindings to a separate `@wordpress/react-data` package. Then the `defaultRegistry` must become part of `@wordpress/data` public API, because the React bindings need to use it.

Taking the above into consideration, I think that `@wordpress/data` should start exporting the `defaultRegistry`. And I like the idea to make it the default export from `@wordpress/data`.

Comment (youknowriad) [reply packages/i18n/src/index.js]: I disagree with exporting the defaultRegistry. I think ideally the defaultRegistry doesn't exist at all and should be instanciated like any other regular registry by the user. The problem though is that we need it for WordPress extensibility and we need it for BC reasons.

Ideally, for me, we should remove the default instances from the packages and add them as inline scripts in WordPress. Package consumers will always use it properly this way while we keep the default instances in WordPress contexts.

Comment (youknowriad) [reply packages/i18n/src/index.js]: The default instances are the ones causing this issue https://github.com/WordPress/gutenberg/issues/8981 so I don't think we should make this worse for package consumers.

Comment (jsnajdr) [reply packages/i18n/src/index.js]: I agree that the default singleton instances shouldn't be there. That applies both to `data` and `i18n`, and also to packages like `core-data` or `block-directory`. These packages statically register a new store in the default registry, instead than exporting constructors and requiring the package consumer to wire up the instances.

But is there a viable and realistic plan to remove the singletons completely? And on what time frame? Or are we stuck with them forever because of backward compatibility?

If the default instances are going to stay, they should receive the love they deserve -- including the `defaultRegistry` export.

Comment (youknowriad) [reply packages/i18n/src/index.js]: Realistically speaking, they are here to stay in WordPress for BC and extensibility reasons but this doesn't mean they should stay in the packages IMO.

Comment (jsnajdr) [reply packages/i18n/src/index.js]: So, the `@wordpress/data` package published in NPM will no longer create and export the default registry (i.e., the `select`, `dispatch` and `registerStore` functions), but when enqueuing the `wp-data` script in WordPress, the `wp.data` global will contain them? So there is no longer a 1:1 correspondence between the `@wordpress/foo` package exports and the `wp.foo` global?

Comment (youknowriad) [reply packages/i18n/src/index.js]: yes and this is already the case for a lot of packages.

Comment (youknowriad) [reply packages/i18n/src/index.js]: See inline scripts here https://github.com/WordPress/WordPress/blob/master/wp-includes/script-loader.php

Comment (jameslnewell) [reply packages/i18n/src/index.js]: I personally prefer the `*` syntax because it doesn't create a bunch of identifiers that have to be kept in sync with another file and it means there's no chance of forgetting to export something.

I've changed the code to suit your preference.

Comment (jameslnewell) [reply packages/i18n/src/index.js]: Is there an outcome to this discussion? How should I proceed?

File: packages/list-reusable-blocks/src/components/import-dropdown/index.js (hunk 10 above)
Comment (jameslnewell) [reply packages/list-reusable-blocks/src/components/import-dropdown/index.js]: Me either!

File: packages/i18n/src/index.js (hunk 2 above)
Comment (jameslnewell) [reply packages/i18n/src/index.js]: 👍That's basically what I had in the first place 😉

File: packages/i18n/src/i18n.js (hunk 6 above)
Comment (jameslnewell) [reply packages/i18n/src/i18n.js]: Have converted the class to a `create*` function.

**Comment by jameslnewell** [comment]
> think it'd also be nice to use this opportunity to refactor the tests added in #20298 to avoid manipulating the require cache and instead create new instances between tests.

@aduth Do you have any opinions around the tests? Do you think we should keep test coverage on both, the `i18n` _instance_ and the default `i18n` _module_ methods or do you think its redundant and we should only test the `i18n` _instance_ and accept the small risk of the `i18n` _module_ methods becoming out-of-sync?

**Comment by aduth** [comment]
I don't think we need to repeat the same tests for the default methods. It should be enough to test methods on an instance created by `createI18n`.

File: packages/i18n/src/index.js (hunk 7 above)
Comment (aduth) [reply packages/i18n/src/index.js]: IMO: Let's not export the default instance. If we need to later, we can add it.

File: packages/i18n/CHANGELOG.md [hunk 11]
```diff
@@ -3,6 +3,7 @@
 ### New Feature
 
 - Add `isRTL` function (#20298)
+- Add `I18n` class to allow creation of multiple i18n instances. (#19210)
```
Comment (sirreal) [inline packages/i18n/CHANGELOG.md]:
```suggestion
- Add `I18n` class to allow creation of multiple i18n instances. (#20318)
```

File: packages/i18n/src/test/sprintf.js [hunk 12]
```diff
@@ -0,0 +1,27 @@
+// Mock memoization as identity function. Inline since Jest errors on out-of-
+// scope references in a mock callback.
```
Comment (sirreal) [inline packages/i18n/src/test/sprintf.js]:
```suggestion
// Mock memoization as identity function. Inline since Jest errors on
// out-of-scope references in a mock callback.
```

File: packages/i18n/src/create-i18n.js [hunk 13]
```diff
@@ -0,0 +1,194 @@
...
+ *																		number, with gettext context.
+ * @property {Function} isRTL 				Check if current locale is RTL.
+ */
+
+/**
+ * Create an i18n instance
+ *
+ * @param {LocaleData} [initialData]   	Locale data configuration.
+ * @param {string}     [initialDomain] 	Domain for which configuration applies.
+ * @return {I18n} 											I18n instance
```
Comment (sirreal) [inline packages/i18n/src/create-i18n.js]: It looks like there's a mix of tabs + spaces for alignment of types/names/descriptions. I believe spaces should be used for consistency.

File: packages/i18n/src/create-i18n.js [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+	 * language written RTL. The opposite of RTL, LTR (Left To Right) is used in other languages,
+	 * including English (`en`, `en-US`, `en-GB`, etc.), Spanish (`es`), and French (`fr`).
+	 *
+	 * @return {boolean} Whether locale is RTL.
+	 */
+	const isRTL = () => {
+		return 'rtl' === _x( 'ltr', 'text direction' );
+	};
+
+	setLocaleData( initialData, initialDomain );
```
Comment (aduth) [inline packages/i18n/src/create-i18n.js]: Should we always be calling this, even if `initialData` and/or `initialDomain` are possibly `undefined` (per the documented types)?

Comment (aduth) [reply packages/i18n/src/create-i18n.js]: Related to this, are there specific use-cases in mind for setting the initial data? It seems like something which _could_ be useful, but just wondering if we might be [getting ahead of ourselves](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it).

Comment (sirreal) [reply packages/i18n/src/create-i18n.js]: I requested this. It seems helpful to be able to populate with initial data, but there's not much lost if we require package consumers to first create then set the data. I don't have strong feelings.

**Review by sirreal — APPROVED** [review]
This works well in my testing.

I don't like that we've moved away from a class which nicely encapsulated state and behavior, but if that's the project convention and preference I'll concede. The end result isn't all that different.

When considering types, now we'll likely end up accessing them as something along the lines of:

```ts
// Create function
type __ = ReturnType< typeof createI18n >['__'];
// Class
type __ = I18n['__']
```

Would we consider using a class internally and exposing a create function that wraps it like was suggested here https://github.com/WordPress/gutenberg/pull/20318#discussion_r383719208

```js
const createI18n = ( ...args ) => new I18n( ...args );
```

**Comment by sirreal** [comment]
I'd like to get this merged, it seems to be in good shape so I'll plan to land it early next week unless there's additional feedback.

cc: @youknowriad @aduth @jsnajdr who've been actively involved in discussion.

**Comment by aduth** [comment]
> Would we consider using a class internally and exposing a create function that wraps it like was suggested here [#20318 (comment)](https://github.com/WordPress/gutenberg/pull/20318#discussion_r383719208)
> 
> ```js
> const createI18n = ( ...args ) => new I18n( ...args );
> ```

Is the idea that the _type_ of the class `I18n` would be available, but not the implementation itself?

File: packages/i18n/src/create-i18n.js (hunk 14 above)
Comment (aduth) [reply packages/i18n/src/create-i18n.js]: It's still odd to me that we call `setLocaleData` if there is nothing to set. It _works_, but only by a combination of the fact that `data` is an optional argument of `setLocaleData` (pretty counter-intuitive), and that object spread is tolerant of undefined (i.e. `{ ...undefined }`), when instead we could just add an `if` here, and make `setLocaleData`'s `data` argument non-optional.

```suggestion
	if ( initialData ) {
		setLocaleData( initialData, initialDomain );
	}
```

Comment (sirreal) [reply packages/i18n/src/create-i18n.js]: @jameslnewell Do you mind addressing this?

**Comment by sirreal** [comment]
> > Would we consider using a class internally and exposing a create function that wraps it like was suggested here [#20318 (comment)](https://github.com/WordPress/gutenberg/pull/20318#discussion_r383719208)
> > ```js
> > const createI18n = ( ...args ) => new I18n( ...args );
> > ```
> 
> Is the idea that the _type_ of the class `I18n` would be available, but not the implementation itself?

I did some testing and this may not be an issue because the type is defined independently. With #18942, it may require this change in the index to export the `I18n` and `LocaleData` types as well:

```js
export * from './create-i18n';
```

We should be able to sort this out after, I'm not concerned about getting this change into this PR.

**Comment by sirreal** [comment]
This appears to be abandoned. I've created https://github.com/WordPress/gutenberg/pull/21182 to address feedback as I can't work on this branch.

File: packages/i18n/src/create-i18n.js (hunk 14 above)
Comment (sirreal) [reply packages/i18n/src/create-i18n.js]: I addressed this in #21182. I believe that's the remaining feedback that needed to be addressed.

**Comment by aduth** [comment]
Closing in favor of #21182.
